go/merge/three_way.go

```diff
@@ -0,0 +1,259 @@
+
+import (
+	"fmt"
+
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/types"
+)
+
+// ErrMergeConflict indicates that a merge attempt failed and must be resolved manually for the provided reason.
+type ErrMergeConflict struct {
+	msg string
+}
+
+func (e *ErrMergeConflict) Error() string {
+	return e.msg
+}
+
+func newMergeConflict(format string, args ...interface{}) *ErrMergeConflict {
+	return &ErrMergeConflict{fmt.Sprintf(format, args...)}
+}
+
+// ThreeWay() attempts a three-way merge between two candidates and a common ancestor. It considers the three of them recursively, applying some simple rules to identify conflicts:
+//  - If any of the three nodes are different types: conflict
+//  - If we are dealing with a map:
+//    - If the same key is both removed and inserted wrt parent: conflict
+//    - If the same key is inserted wrt parent, but with different values: conflict
+//  - If we are dealing with a struct:
+//    - If the same field is both removed and inserted wrt parent: conflict
+//    - If the same field is inserted wrt parent, but with different values: conflict
+//  - If we are dealing with a list:
+//    - If the same index is both removed and inserted wrt parent: conflict
+//    - If the same index is inserted wrt parent, but with different values: conflict
+//  - If we are dealing with a set:
+//    - If the same object is both removed and inserted wrt parent: conflict
+//
+// All other modifications are allowed.
+// Currently, ThreeWay() only works on types.Map.
+func ThreeWay(a, b, parent types.Value, vwr types.ValueReadWriter) (merged types.Value, err error) {
+	if a == nil && b == nil {
+		return parent, nil
+	} else if a == nil {
+		return parent, newMergeConflict("Cannot merge nil Value with %s.", b.Type().Describe())
+	} else if b == nil {
+		return parent, newMergeConflict("Cannot merge %s with nil value.", a.Type().Describe())
+	} else if unmergeable(a, b) {
+		return parent, newMergeConflict("Cannot merge %s with %s.", a.Type().Describe(), b.Type().Describe())
+	}
+
+	return threeWayMerge(a, b, parent, vwr)
+}
+
+// a and b cannot be merged if they are of different NomsKind, or if at least one of the two is nil, or if either is a Noms primitive.
+func unmergeable(a, b types.Value) bool {
+	if a != nil && b != nil {
+		aKind, bKind := a.Type().Kind(), b.Type().Kind()
+		return aKind != bKind || types.IsPrimitiveKind(aKind) || types.IsPrimitiveKind(bKind)
+	}
+	return true
+}
+
+func threeWayMerge(a, b, parent types.Value, vwr types.ValueReadWriter) (merged types.Value, err error) {
+	d.PanicIfTrue(a == nil || b == nil, "Merge candidates cannont be nil: a = %v, b = %v", a, b)
+	newTypeConflict := func() *ErrMergeConflict {
+		pDescription := "<nil>"
+		if parent != nil {
+			pDescription = parent.Type().Describe()
+		}
+		return newMergeConflict("Cannot merge %s and %s on top of %s.", a.Type().Describe(), b.Type().Describe(), pDescription)
+	}
+
+	switch a.Type().Kind() {
+	default:
```

Comment:
I like to do this if `default` is the common case, but that's not really the case here, so I've changed it.